src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -1575,24 +1575,10 @@ public DecisionDag(DagState rootNode)
             /// <summary>
             /// A successor function used to topologically sort the DagState set.
             /// </summary>
-            private static ImmutableArray<DagState> Successor(DagState state)
+            private static void AddSuccessor(ArrayBuilder<DagState> builder, DagState state)
             {
-                if (state.TrueBranch != null && state.FalseBranch != null)
-                {
-                    return ImmutableArray.Create(state.FalseBranch, state.TrueBranch);
-                }
-                else if (state.TrueBranch != null)
-                {
-                    return ImmutableArray.Create(state.TrueBranch);
-                }
-                else if (state.FalseBranch != null)
-                {
-                    return ImmutableArray.Create(state.FalseBranch);
-                }
-                else
-                {
-                    return ImmutableArray<DagState>.Empty;
-                }
+                builder.AddIfNotNull(state.TrueBranch);
+                builder.AddIfNotNull(state.FalseBranch);
```

Comment:
this guy was particularly egregious.  Constantly created arrays of just 1-2 elements long just to then walk and throw them away.  

These algorithms really benefit from just using a scratch buffer to write things into.  

If this wasn't a generic algorithm, and was known to be used *only* by the decisinon-dag stuff, we could optimize even further by just using a stack allocated span for this.  but i didn't want to go that far.

Reply:
Can we not use TemporaryArray<T> here? It has better cache performance than the object pools when the majority of cases do not exceed 4 elements.

Reply:
these take in delegates that are called to add elements to the scratch space.  We'd need to pass temporary-array by-ref to make that work, and i was hesitant about defining a whole new delegate type for that.  i'm not opposed to it, but this was def path of least resistance.  And the builders themselves are only created once for the lifetime of the whole graph walk, so this seemed good enough (For now).

But i am open to changing this in this PR if we feel TempArray would be ncier.